flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+					String newState = function.apply(state, delta);
+					keyEntry.setValue(newState);
+					stateMap.transform(key, namespace, delta, function);
+					assertEquals(totalSize, stateMap.size());
+					assertEquals(totalSize, stateMap.totalSize());
+				}
+			}
+		}
+
+		// transform some new states
+		for (long namespace = 15; namespace < 20; namespace++) {
+			for (int key = 0; key < 100; key++) {
+				totalSize++;
+				int value = (int) (key * namespace);
+				stateMap.transform(key, namespace, value, function);
+				referenceStates.computeIfAbsent(namespace, (none) -> new HashMap<>()).put(key, String.valueOf(value));
+				assertEquals(totalSize, stateMap.size());
+				assertEquals(totalSize, stateMap.totalSize());
+			}
+		}
+
+		assertEquals(totalSize * 2, spaceAllocator.getTotalSpaceNumber());
+		verifyState(referenceStates, stateMap);
+
+		// put some previously removed states
+		for (Map.Entry<Long, Set<Integer>> entry : removedStates.entrySet()) {
+			long namespace = entry.getKey();
+			for (int key : entry.getValue()) {
+				totalSize++;
+				String state = String.valueOf(key * namespace);
+				if (random.nextBoolean()) {
+					stateMap.put(key, namespace, state);
+				} else {
+					assertNull(stateMap.putAndGetOld(key, namespace, state));
+				}
+				referenceStates.computeIfAbsent(namespace, (none) -> new HashMap<>()).put(key, String.valueOf(state));
+				assertEquals(totalSize, stateMap.size());
+				assertEquals(totalSize, stateMap.totalSize());
+			}
+		}
+
+		assertEquals(totalSize * 2, spaceAllocator.getTotalSpaceNumber());
+		verifyState(referenceStates, stateMap);
+
+		// remove some namespaces
+		Set<Long> removedNamespaces = new HashSet<>();
+		for (Map.Entry<Long, Map<Integer, String>> namespaceEntry : referenceStates.entrySet()) {
+			if (random.nextBoolean()) {
+				long namespace = namespaceEntry.getKey();
+				removedNamespaces.add(namespace);
+				for (Map.Entry<Integer, String> keyEntry : namespaceEntry.getValue().entrySet()) {
+					int key = keyEntry.getKey();
+					if (random.nextBoolean()) {
+						stateMap.remove(key, namespace);
+					} else {
+						assertEquals(keyEntry.getValue(), stateMap.removeAndGetOld(key, namespace));
+					}
+					totalSize--;
+					assertEquals(totalSize, stateMap.size());
+					assertEquals(totalSize, stateMap.totalSize());
+				}
+			}
+		}
+
+		for (long namespace : removedNamespaces) {
+			referenceStates.remove(namespace);
+			assertEquals(0, stateMap.sizeOfNamespace(namespace));
+			assertFalse(stateMap.getKeys(namespace).iterator().hasNext());
+		}
+
+		assertEquals(totalSize * 2, spaceAllocator.getTotalSpaceNumber());
+		verifyState(referenceStates, stateMap);
+
+		stateMap.close();
+		assertEquals(0, stateMap.size());
+		assertEquals(0, stateMap.totalSize());
+		// all spaces should be free
+		assertEquals(0, spaceAllocator.getTotalSpaceNumber());
+		assertTrue(stateMap.isClosed());
+	}
+
+	/**
+	 *  Tests copy-on-write contracts.
+	 */
+	@SuppressWarnings("unchecked")
```

Comment:
I think this annotation is not needed if `#stateSnapshot` would return a `CopyOnWriteSkipListStateMapSnapshot`.